pageserver/src/layered_repository.rs

```diff
@@ -1938,41 +1938,71 @@ impl LayeredTimeline {
         Ok(new_path)
     }
 
+    ///
+    /// Collect a bunch of Level 0 layer files, and compact and reshuffle them as
+    /// as Level 1 files.
+    ///
     fn compact_level0(&self, target_file_size: u64) -> Result<()> {
         let layers = self.layers.read().unwrap();
-
-        let level0_deltas = layers.get_level0_deltas()?;
+        let mut level0_deltas = layers.get_level0_deltas()?;
+        drop(layers);
 
         // We compact or "shuffle" the level-0 delta layers when they've
         // accumulated over the compaction threshold.
-        if level0_deltas.len() < self.get_compaction_threshold() {
+        if level0_deltas.is_empty() || level0_deltas.len() < self.get_compaction_threshold() {
             return Ok(());
         }
-        drop(layers);
 
-        // FIXME: this function probably won't work correctly if there's overlap
-        // in the deltas.
-        let lsn_range = level0_deltas
-            .iter()
-            .map(|l| l.get_lsn_range())
-            .reduce(|a, b| min(a.start, b.start)..max(a.end, b.end))
-            .unwrap();
+        level0_deltas.sort_by_key(|l| l.get_lsn_range().start);
+        let mut level0_deltas_iter = level0_deltas.iter();
+
+        let first_level0_delta = level0_deltas_iter.next().unwrap();
+        let mut prev_lsn_end = first_level0_delta.get_lsn_range().end;
+        let mut deltas_to_compact = vec![Arc::clone(first_level0_delta)];
+        for l in level0_deltas_iter {
+            let lsn_range = l.get_lsn_range();
+
+            if lsn_range.start != prev_lsn_end {
+                break;
+            }
+            deltas_to_compact.push(Arc::clone(l));
+            prev_lsn_end = lsn_range.end;
+        }
+        let lsn_range = Range {
+            start: deltas_to_compact.first().unwrap().get_lsn_range().start,
+            end: deltas_to_compact.last().unwrap().get_lsn_range().end,
+        };
+
+        info!(
+            "Starting Level0 compaction in LSN range {}-{} for {} layers ({} deltas in total)",
+            lsn_range.start,
+            lsn_range.end,
+            deltas_to_compact.len(),
+            level0_deltas.len()
+        );
+        for l in deltas_to_compact.iter() {
+            info!("compact includes {}", l.filename().display());
+        }
+        drop(level0_deltas);
```

Comment:
Nit: the above code can be rearranged, so that we can hide the mutability of the `deltas_to_compact` vector and don't need to explicitly drop `level0_deltas` and `layers`. But I guess that would require not trivial changes

Reply:
Hmm, I agree that would be nice, but I also don't see a nice way to structure that. I added a comment explaining that the point of the 'drop' is to make sure you don't accidentally use it later in the function.